notify_socket.go

```diff
@@ -142,29 +145,86 @@ func (n *notifySocket) run(pid1 int) error {
 				return nil
 			}
 		case b := <-fileChan:
-			var out bytes.Buffer
-			_, err = out.Write(b)
-			if err != nil {
-				return err
-			}
+			return notifyHost(client, b, pid1)
+		}
+	}
+}
 
-			_, err = out.Write([]byte{'\n'})
-			if err != nil {
-				return err
-			}
+// notifyHost tells the host (usually systemd) that the container reported READY.
+// Also sends MAINPID and BARRIER.
+func notifyHost(client *net.UnixConn, ready []byte, pid1 int) error {
+	var out bytes.Buffer
+	_, err := out.Write(ready)
+	if err != nil {
+		return err
+	}
 
-			_, err = client.Write(out.Bytes())
-			if err != nil {
-				return err
-			}
+	_, err = out.Write([]byte{'\n'})
+	if err != nil {
+		return err
+	}
 
-			// now we can inform systemd to use pid1 as the pid to monitor
-			newPid := "MAINPID=" + strconv.Itoa(pid1)
-			_, err := client.Write([]byte(newPid + "\n"))
-			if err != nil {
-				return err
-			}
-			return nil
-		}
+	_, err = client.Write(out.Bytes())
```

Comment:
Hmm, wonder why we avoid using bytes.Buffer at all and don't just do something like
```go
_, err = client.WriteString(string(ready)+"\n")
```

Reply:
> Hmm, wonder why we avoid using bytes.Buffer at all 

Looking at git history, this was done in #1308 but I don't see why.

Reply:
> _, err = client.Write(append(ready, '\n'))

This one looks fine to me (most probably there's `\n` already anyway at that position), as we are not reusing the buffer anywhere.

Can you implement it (as a separate first commit, I guess)?